### PR TITLE
[VL] Refactor MemoryManager#getArrowMemoryPool to return raw pointer

### PR DESCRIPTION
**Pull request by Yohahaha**
## What changes were proposed in this pull request?

Fix exist TODO, caller was not expected to care memory pool's ownership, just get raw pointer for use, like Arrow style.

### CODE REVIEW

File: cpp/velox/shuffle/VeloxShuffleWriter.cc
Comment (Yohahaha) [inline cpp/velox/shuffle/VeloxShuffleWriter.cc:408]: Changes here are tricky, current codes set ipc_memory_pool with memory_pool in JniWrapper.cc#ShuffleWriterJniWrapper_nativeMake, after check several related patches I assume LargeMemoryPool was not used.

We may direct use same arrow pool for shuffle writer and provide a config to decide whether to use LargeMemoryPool when initialize it in MemoryManager.

@marin-ma could you help review this change? thanks!

Comment (marin-ma) [reply cpp/velox/shuffle/VeloxShuffleWriter.cc:408]: Thanks for your advice. Let's add a configuration for it.

Some background: LargeMemoryPool was introduced to alleviate the occurrence of page fault, and it was only used for shuffle compression. But since the shuffle compression part has been refactored many times, we haven't done performance testing around the latest implementation, so it's not used for now.

Comment (marin-ma) [reply cpp/velox/shuffle/VeloxShuffleWriter.cc:408]: cc: @FelixYBW

Comment (zhztheplayer) [reply cpp/velox/shuffle/VeloxShuffleWriter.cc:408]: I suggest avoiding bringing this sort of change to the final version of the PR. Storing a static instance is an option if it's not in use but we don't decide to remove it either? @Yohahaha

Comment (Yohahaha) [reply cpp/velox/shuffle/VeloxShuffleWriter.cc:408]: yeah, this is not the final version, I prefer to add a configuration for using LargeMemoryPool first, and then finish this patch.

Comment (Yohahaha) [reply cpp/velox/shuffle/VeloxShuffleWriter.cc]: > I prefer to add a configuration for using LargeMemoryPool first

After take some time to read codes, I give up it because add an unused config just for keep unused code does not make sense for me, and it will bring lots of changes in VeloxMemoryManager.

I keep codes in VeloxShuffleWriter.cc but refactor a little, please help review again, thanks! @marin-ma @zhztheplayer

**Comment by Yohahaha** [comment]
could anyone help rerun failed jobs?

**Comment by zhztheplayer** [comment]
> could anyone help rerun failed jobs?

Rerunning doesn't fix the issue, would you check the job logs? Thanks.

**Comment by zhztheplayer** [comment]
Lots of logs like following:

```
2023-09-14T06:07:31.6337350Z 23/09/14 14:07:31 ERROR ShuffleClientImpl: Push data to 172.17.0.4:42863 failed for shuffle 0 map 14 attempt 0 partition 69 batch 70, remain revive times 5.
2023-09-14T06:07:31.6338836Z java.io.IOException: Connection from /172.17.0.4:42863 closed
2023-09-14T06:07:31.6345650Z 	at org.apache.celeborn.common.network.client.TransportResponseHandler.channelInactive(TransportResponseHandler.java:241)
2023-09-14T06:07:31.6350645Z 	at org.apache.celeborn.common.network.server.TransportChannelHandler.channelInactive(TransportChannelHandler.java:141)
2023-09-14T06:07:31.6355853Z 	at org.apache.celeborn.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:305)
2023-09-14T06:07:31.6361283Z 	at org.apache.celeborn.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:281)
2023-09-14T06:07:31.6366995Z 	at org.apache.celeborn.shaded.io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:274)
2023-09-14T06:07:31.6372258Z 	at org.apache.celeborn.shaded.io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:81)
2023-09-14T06:07:31.6377190Z 	at org.apache.celeborn.shaded.io.netty.handler.timeout.IdleStateHandler.channelInactive(IdleStateHandler.java:277)
2023-09-14T06:07:31.6382274Z 	at org.apache.celeborn.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:303)
2023-09-14T06:07:31.6388376Z 	at org.apache.celeborn.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:281)
2023-09-14T06:07:31.6394624Z 	at org.apache.celeborn.shaded.io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:274)
2023-09-14T06:07:31.6400092Z 	at org.apache.celeborn.shaded.io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:81)
2023-09-14T06:07:31.6404163Z 	at org.apache.celeborn.common.network.util.TransportFrameDecoder.channelInactive(TransportFrameDecoder.java:175)
2023-09-14T06:07:31.6410760Z 	at org.apache.celeborn.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:305)
2023-09-14T06:07:31.6416076Z 	at org.apache.celeborn.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:281)
2023-09-14T06:07:31.6421653Z 	at org.apache.celeborn.shaded.io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:274)
2023-09-14T06:07:31.6426678Z 	at org.apache.celeborn.shaded.io.netty.channel.DefaultChannelPipeline$HeadContext.channelInactive(DefaultChannelPipeline.java:1405)
2023-09-14T06:07:31.6431971Z 	at org.apache.celeborn.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:301)
2023-09-14T06:07:31.6437147Z 	at org.apache.celeborn.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:281)
2023-09-14T06:07:31.6442672Z 	at org.apache.celeborn.shaded.io.netty.channel.DefaultChannelPipeline.fireChannelInactive(DefaultChannelPipeline.java:901)
2023-09-14T06:07:31.6448059Z 	at org.apache.celeborn.shaded.io.netty.channel.AbstractChannel$AbstractUnsafe$7.run(AbstractChannel.java:813)
2023-09-14T06:07:31.6452107Z 	at org.apache.celeborn.shaded.io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
2023-09-14T06:07:31.6456804Z 	at org.apache.celeborn.shaded.io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167)
2023-09-14T06:07:31.6461135Z 	at org.apache.celeborn.shaded.io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
2023-09-14T06:07:31.6465605Z 	at org.apache.celeborn.shaded.io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
2023-09-14T06:07:31.6469400Z 	at org.apache.celeborn.shaded.io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
2023-09-14T06:07:31.6474130Z 	at org.apache.celeborn.shaded.io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
2023-09-14T06:07:31.6477985Z 	at org.apache.celeborn.shaded.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
2023-09-14T06:07:31.6481007Z 	at java.lang.Thread.run(Thread.java:750)
```

**Comment by Yohahaha** [comment]
CI has passed, could you help review again? thanks! @zhztheplayer @marin-ma

File: cpp/velox/shuffle/VeloxShuffleWriter.cc
Comment (zhztheplayer) [inline cpp/velox/shuffle/VeloxShuffleWriter.cc:597]: Why not just leave a smart pointer for options_.ipc_memory_pool since it's just test code? `delete` still looks risky here.

Besides, in previous code why the if condition was needed?

```cpp
if (options_.ipc_memory_pool != options_.memory_pool) {
  options_.ipc_memory_pool.reset();
}
```

Why can't we use
```cpp
options_.ipc_memory_pool.reset();
```
instead? @marin-ma

Comment (Yohahaha) [reply cpp/velox/shuffle/VeloxShuffleWriter.cc:597]: > Why not just leave a smart pointer for options_.ipc_memory_pool since it's just test code?

Two solutions here:
1. keep options_.ipc_memory_pool but comments its usage at `VeloxShuffleWriter#initIpcWriteOptions`, since these codes are not used. 
2. current implementation.

The key is we need set ipc_memory_pool with memory_pool to enable this unused in-test feature, keep these two pool with same type is better.

**Comment by Yohahaha** [comment]
pending this patch until #3159 merged.

**Comment by Yohahaha** [comment]
CI may related with #3235

**Review by zhztheplayer — COMMENTED** [review]
LGTM @marin-ma If you have any other comments.

File: cpp/velox/shuffle/VeloxShuffleWriter.cc
Comment (marin-ma) [inline cpp/velox/shuffle/VeloxShuffleWriter.cc:1552]: This is an unexpected change. We had shuffle writer memory pool separated into payloadPool_ and partitionBufferPool_, to help count the allocation of separately. Please refer to #3159 .

File: cpp/velox/shuffle/VeloxShuffleWriter.cc
Comment (marin-ma) [inline cpp/velox/shuffle/VeloxShuffleWriter.cc:1559]: Same as above.

File: cpp/velox/shuffle/VeloxShuffleWriter.cc
Comment (Yohahaha) [reply cpp/velox/shuffle/VeloxShuffleWriter.cc]: Thanks, already fixed.

**Review by marin-ma — APPROVED** [review]

**Comment by marin-ma** [comment]
> ===== Performance report for TPCH SF2000 with Velox backend, for reference only ====
> 
> query	log/native_3120_time.csv	log/native_master_09_27_2023_523395c22_time.csv	difference	percentage
> q1	43.33	43.84	0.507	101.17%
> q2	26.96	24.25	-2.707	89.96%
> q3	40.97	37.31	-3.661	91.06%
> q4	44.94	41.10	-3.845	91.44%
> q5	73.21	70.67	-2.541	96.53%
> q6	7.11	7.07	-0.034	99.52%
> q7	89.81	83.62	-6.193	93.10%
> q8	84.37	82.10	-2.265	97.32%
> q9	121.47	116.36	-5.107	95.80%
> q10	49.93	46.38	-3.555	92.88%
> q11	20.83	18.95	-1.882	90.97%
> q12	22.84	25.24	2.405	110.53%
> q13	52.73	50.58	-2.145	95.93%
> q14	16.58	16.28	-0.299	98.20%
> q15	32.03	27.18	-4.856	84.84%
> q16	16.60	16.28	-0.324	98.05%
> q17	131.56	120.13	-11.428	91.31%
> q18	166.76	161.84	-4.919	97.05%
> q19	12.79	12.16	-0.628	95.09%
> q20	38.46	28.31	-10.142	73.63%
> q21	264.12	239.00	-25.125	90.49%
> q22	16.17	15.72	-0.448	97.23%
> total	1373.55	1284.35	-89.192	93.51%

Why performance dropped so much?

**Pull request merged**
